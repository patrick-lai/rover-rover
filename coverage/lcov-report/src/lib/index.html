<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/lib</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> src/lib
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">88.44% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>130/147</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">87.23% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>82/94</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">96.3% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>26/27</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">90.55% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>115/127</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line high'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="Plateau.js"><a href="Plateau.js.html">Plateau.js</a></td>
	<td data-value="87.01" class="pic high"><div class="chart"><div class="cover-fill" style="width: 87%;"></div><div class="cover-empty" style="width:13%;"></div></div></td>
	<td data-value="87.01" class="pct high">87.01%</td>
	<td data-value="77" class="abs high">67/77</td>
	<td data-value="86.84" class="pct high">86.84%</td>
	<td data-value="38" class="abs high">33/38</td>
	<td data-value="94.12" class="pct high">94.12%</td>
	<td data-value="17" class="abs high">16/17</td>
	<td data-value="88.57" class="pct high">88.57%</td>
	<td data-value="70" class="abs high">62/70</td>
	</tr>

<tr>
	<td class="file high" data-value="Rover.js"><a href="Rover.js.html">Rover.js</a></td>
	<td data-value="85.11" class="pic high"><div class="chart"><div class="cover-fill" style="width: 85%;"></div><div class="cover-empty" style="width:15%;"></div></div></td>
	<td data-value="85.11" class="pct high">85.11%</td>
	<td data-value="47" class="abs high">40/47</td>
	<td data-value="82.5" class="pct high">82.5%</td>
	<td data-value="40" class="abs high">33/40</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="90.24" class="pct high">90.24%</td>
	<td data-value="41" class="abs high">37/41</td>
	</tr>

<tr>
	<td class="file high" data-value="instructions.js"><a href="instructions.js.html">instructions.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="23" class="abs high">23/23</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="16" class="abs high">16/16</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="16" class="abs high">16/16</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Fri Mar 23 2018 01:09:46 GMT+1100 (AEDT)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
