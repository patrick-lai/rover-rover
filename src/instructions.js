/***
 * Definitions of instructions and some helpers
 */

export const DIRECTIONS = ['N', 'E', 'S', 'W'];
export const INSTRUCTIONS = ['L', 'R', 'M'];

// Expected Format: Int Int Direction
export const isValidLandingSpot = input => {
  if (typeof input !== 'string') return false;
  // Lets just ignore extra instructions
  const landingSpot = input.split(' ');
  if (isNaN(parseInt(landingSpot[0], 10))) return false;
  if (isNaN(parseInt(landingSpot[1], 10))) return false;
  if (!DIRECTIONS.includes(landingSpot[2])) return false;
  return true;
};

// Just check if each instruction is valid
export const isValidInstructions = instructions => {
  if (typeof instructions !== 'string') return false;
  const _instructions = instructions.split('');
  for (const i of _instructions) if (!INSTRUCTIONS.includes(i)) return false;
  return true;
};
