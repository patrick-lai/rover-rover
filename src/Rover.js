/***
 * A simple rover
 */

import { isValidLandingSpot, isValidInstructions, DIRECTIONS, INSTRUCTIONS } from './instructions';

export default class Rover {
  // We want to disable the rover if it tries to do anything unexpected
  disabled = false;
  // Currently facing direction
  facing = null;
  // List of seen directions
  seen = [];
  // Some warning messages
  warnings = [];

  constructor(config = {}) {
    const { instructions, landingSpot } = config;
    if (!isValidLandingSpot(landingSpot)) throw new Error('Invalid landing spot');
    if (!isValidInstructions(instructions)) this.warnings.push['Invalid instructions'];

    // Setup data structures (Array)
    this.landingSpot = landingSpot.split(' ');
    this.instructions = instructions && instructions.split('');

    // Check config
    if (!landingSpot) throw new Error('Rover cannot be constructed without a landingSpot');
  }

  /*
   * Utilities
   */

  _isCompleted = () => this.instructions.length === 0;

  /*
   * Public methods
   */

  // Initiates a rover and sets its initial facing position
  // useful when we want to control when to land it
  init = () => this.setFacing(this.landingSpot[2]);

  // We purposely wont have a method to enable it
  disableRover = () => {
    this.disabled = true;
  };

  // Adds a direction to seen and set new facing direction
  setFacing = direction => {
    const { seen } = this;
    if (!DIRECTIONS.includes(direction)) throw new Error('Invalid direction');
    if (!seen.includes(direction)) seen.push(direction);
    this.facing = direction;
    return this.facing;
  };

  // Checks if the rover has seen a list of directions
  hasSeen = directions => {
    if (!Array.isArray(directions)) throw new Error('Expected array as an input for hasSeen');
    for (const d of directions) if (!this.seen.includes(d)) return false;
    return true;
  };

  executeNextInstruction = () => {
    if (this.facing === null) throw new Error('Rover has not be landed yet');
    if (this.disabled) return console.warn('Rover has been disabled, failed to execute instruction');
    if (this._isCompleted()) return console.warn('Tried to execute an empty instruction set');

    // NOTE - we can look to save it in an executed log but unneccesary for now
    const instruction = this.instructions.shift();

    // Used to calculate new direction
    const index = DIRECTIONS.indexOf(this.facing);
    let newDirection;

    switch (instruction) {
      case 'L':
        newDirection = DIRECTIONS[index > 0 ? index - 1 : DIRECTIONS.length - 1];
        break;
      case 'R':
        newDirection = DIRECTIONS[index < DIRECTIONS.length - 1 ? index + 1 : 0];
        break;
      case 'M':
        // Do nothing the plateau handles this
        break;
      default:
        throw new Error(`Unexpected instruction ${instruction}`);
    }

    // Update the new direction
    if (newDirection) this.setFacing(newDirection);

    return instruction;
  };
}
