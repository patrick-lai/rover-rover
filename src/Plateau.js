/***
 * Generate an n*n iterable plateau
 * Handles rover movement
 */

/*
 * Helpers
 */

import Rover from './Rover';

const validateMaxStackingRovers = input => input > 0;

const validateDimensions = input => {
  // Lets just ignore extra instructions
  const dimensions = input.split(' ');

  const width = parseInt(dimensions[0], 10);
  const height = parseInt(dimensions[1], 10);

  if (isNaN(width) || width < 1) return false;
  if (isNaN(height) || height < 1) return false;
  return true;
};

export default class Plateau {
  rovers = [];
  maxStackingRovers;
  constructor(config = {}) {
    const { maxStackingRovers, dimensions } = config;

    if (!validateDimensions(dimensions)) throw new Error('Invalid dimensions, must be 2 Ints spaced');
    if (!validateMaxStackingRovers(maxStackingRovers))
      throw new Error('Invalid maxStackingRovers, must be an integer larger than 0');

    this.maxStackingRovers = maxStackingRovers;
  }

  deployRover = rover => {
    if (rover.constructor !== Rover) throw new Error('Can only add type Rover');
    this.rovers.push(rover);

    // Lets land it as per spec
    rover.init();
  };
}
