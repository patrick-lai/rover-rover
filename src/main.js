/***
 * From an input string execute some rovers on a plateau
 */

require('dotenv').config();
import Plateau from './lib/Plateau';
import Rover from './lib/Rover';
import { isValidLandingSpot, isValidInstructions } from './lib/instructions';

const filename = process.argv[2];
const { MAX_STACKING_ROVERS } = process.env;
const maxStackingRovers = parseInt(MAX_STACKING_ROVERS, 10);
if (isNaN(maxStackingRovers))
  throw new Error('Expected MAX_STACKING_ROVERS to be an Integer in env');

export const executeInput = data => {
  const rovers = [];
  // Split the data by lines, filter out trailing lines/spaces
  const rows = data.split('\n').filter(r => r && r !== '');

  // Consume first row
  const plateau = new Plateau({ maxStackingRovers, dimensions: rows.shift() });

  // Make some rovers
  while (rows.length > 0) {
    const landingSpot = rows.shift();
    const instructions = rows.shift();

    if (!isValidLandingSpot(landingSpot))
      throw new Error(`Invalid landing spot: ${landingSpot}`);

    if (!isValidInstructions(instructions))
      throw new Error(`Invalid instructions: ${instructions}`);

    const rover = new Rover({ landingSpot, instructions });
    plateau.deployRover(rover);
  }

  // Execute all rovers
  plateau.rovers.forEach(rover => {
    while (!rover.isCompleted()) plateau.executeNextInstructionOfRover(rover);
  });

  const roverStates = plateau.rovers.map(rover => plateau.getRoverState(rover));
  return roverStates.join('\n');
};
