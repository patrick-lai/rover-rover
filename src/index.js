/***
 * Executes some text input for rovers on a plateau
 */

import fs from 'fs';
import { executeInput } from './main';

const filename = process.argv[2];
const { MAX_STACKING_ROVERS } = process.env;
const maxStackingRovers = parseInt(MAX_STACKING_ROVERS, 10);

// Checks
if (!filename) throw new Error('Please specify a filename');
if (isNaN(maxStackingRovers))
  throw new Error('Expected MAX_STACKING_ROVERS to be an Integer in env');

// Execute
fs.readFile(filename, 'utf8', (err, input) => {
  if (err) throw err;
  console.info(`Max rovers allowed at one point: ${maxStackingRovers}\n`);
  console.info(`Final rover states\n`);
  console.log(executeInput(input));
});
