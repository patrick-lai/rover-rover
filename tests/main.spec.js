import fs from 'fs';
import { executeInput } from '../src/main';

// OVERRIDE ENV
process.env.MAX_STACKING_ROVERS = 1;

describe('Rover behaviour', () => {
  test('Can run test case', done => {
    const SAMPLE_FILE = './tests/sampleData/sampleInput.txt';
    fs.readFile(SAMPLE_FILE, 'utf8', (err, input) => {
      if (err) throw err;
      expect(executeInput(input)).toMatchSnapshot();
      done();
    });
  });

  test('Can run large complex case', done => {
    const SAMPLE_FILE = './tests/sampleData/largeGrid.txt';
    fs.readFile(SAMPLE_FILE, 'utf8', (err, input) => {
      if (err) throw err;
      expect(executeInput(input)).toMatchSnapshot();
      done();
    });
  });

  test('One failing doesnt stop the whole fleet', done => {
    const SAMPLE_FILE = './tests/sampleData/crash.txt';
    fs.readFile(SAMPLE_FILE, 'utf8', (err, input) => {
      if (err) throw err;
      expect(executeInput(input)).toMatchSnapshot();
      done();
    });
  });

  test('Rover stops before it goes off the plateau', done => {
    const SAMPLE_FILE = './tests/sampleData/outOfBounds.txt';
    fs.readFile(SAMPLE_FILE, 'utf8', (err, input) => {
      if (err) throw err;
      expect(executeInput(input)).toMatchSnapshot();
      done();
    });
  });
});
