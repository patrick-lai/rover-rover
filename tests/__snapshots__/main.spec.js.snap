// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rover behaviour Can run large complex case 1`] = `
"52 53 N
30 42 W
20 21 E
19 9 W
84 24 N
27 22 S"
`;

exports[`Rover behaviour Can run test case 1`] = `
"1 3 N
5 1 E"
`;

exports[`Rover behaviour One failing doesnt stop the whole fleet 1`] = `
"1 3 W
1 3 N
1 3 N
1 2 N
1 2 N
5 1 E
2 4 N"
`;

exports[`Rover behaviour Rover stops before it goes off the plateau 1`] = `
"1 3 N
3 2 E
3 3 E
0 1 W"
`;
